core/service/src/bin/kms-server.rs

```diff
@@ -193,109 +205,84 @@ async fn build_tls_config(
-                &PubDataType::CACert.to_string(),
-            )
-            .await?;
-            let ca_cert = x509_parser::pem::parse_x509_pem(ca_cert_bytes.as_bytes())?.1;
-
-            // check if the CA certificate matches the KMS signing key
-            let ca_cert_x509 = ca_cert.parse_x509()?;
-            if let x509_parser::public_key::PublicKey::EC(pk_sec1) =
-                ca_cert_x509.public_key().parsed()?
-            {
-                let ca_pk = Box::new(pk_sec1.data());
-                #[allow(deprecated)]
-                let sk_vk = sk.sk().verifying_key().to_encoded_point(false).to_bytes();
-                ensure!(
+            let (sk, ca_cert) = match eif_signing_cert {
+                Some(eif_signing_cert) => {
+                    tracing::info!("Using wrapped TLS certificate with Nitro remote attestation");
+                    (
+                        None,
+                        match peers {
+                            Some(ref peers) => eif_signing_cert.into_pem(my_id, peers)?,
+                            None => {
+                                tracing::info!(
+                        "No peerlist present, skipping TLS certificate validation against peerlist"
+                    );
+                                eif_signing_cert.unchecked_pem()?
+                            }
+                        },
+                    )
+                }
+                None => {
+                    tracing::info!(
+                        "Using TLS certificate with Nitro remote attestation signed by onboard CA"
+                    );
+                    let ca_cert_bytes = read_text_at_request_id(
+                        public_vault,
+                        &SIGNING_KEY_ID,
+                        &PubDataType::CACert.to_string(),
+                    )
+                    .await?;
+                    let ca_cert = x509_parser::pem::parse_x509_pem(ca_cert_bytes.as_bytes())?.1;
+
+                    // check if the CA certificate matches the KMS signing key
+                    let ca_cert_x509 = ca_cert.parse_x509()?;
+                    if let x509_parser::public_key::PublicKey::EC(pk_sec1) =
+                        ca_cert_x509.public_key().parsed()?
+                    {
+                        let ca_pk = Box::new(pk_sec1.data());
+                        #[allow(deprecated)]
+                        let sk_vk = sk.sk().verifying_key().to_encoded_point(false).to_bytes();
+                        ensure!(
                     **ca_pk == *sk_vk,
                     "CA certificate public key {:?} doesn't correspond to the KMS verifying key {:?}",
                     hex::encode(*ca_pk),
                     hex::encode(sk_vk)
-                );
-            } else {
-                panic!("CA certificate public key isn't ECDSA");
+                        );
+                    } else {
+                        panic!("CA certificate public key isn't ECDSA");
```

Comment:
I'm assuming it's OK to panic here, cause this happens very early in the process. Or should we rather have Error handling that we can handle on the caller side?